src/analyser.cpp

```diff
@@ -1349,129 +1373,322 @@ void Analyser::AnalyserImpl::analyseModel(const ModelPtr &model)
 
     if ((mModel->mPimpl->mType == AnalyserModel::Type::ODE)
         || (mModel->mPimpl->mType == AnalyserModel::Type::ALGEBRAIC)) {
-        // Sort our internal variables and equations.
+        // Add a dummy equation for each of our true (i.e. non-computed)
+        // constants.
+        // Note: this is only so that we can mark a constant as an external
+        //       variable.
 
-        std::sort(mInternalVariables.begin(), mInternalVariables.end(),
-                  compareVariablesByTypeAndIndex);
-        std::sort(mInternalEquations.begin(), mInternalEquations.end(),
-                  compareEquationsByVariable);
+        for (const auto &internalVariable : mInternalVariables) {
+            if (internalVariable->mType == AnalyserInternalVariable::Type::CONSTANT) {
+                mInternalEquations.push_back(std::shared_ptr<AnalyserInternalEquation> {new AnalyserInternalEquation {internalVariable}});
+            }
+        }
 
-        std::map<VariablePtr, AnalyserEquationPtr> equationMappings;
-        std::map<AnalyserEquationPtr, AnalyserVariablePtr> variableMappings;
+        // Mark some variables as external variables, if needed.
 
-        for (const auto &internalEquation : mInternalEquations) {
-            equationMappings[internalEquation->mVariable->mVariable] = std::shared_ptr<AnalyserEquation> {new AnalyserEquation {}};
-        }
+        std::map<AnalyserInternalVariablePtr, std::vector<VariablePtr>> externalVariables;
 
-        // Make our internal variables available through our API.
+        if (!mExternalVariables.empty()) {
+            // Check whether an external variable belongs to the model being
+            // analysed, or whether it is marked as an external variable more
+            // than once through equivalence or is (equivalent to) the variable
+            // of integration.
 
-        auto stateIndex = MAX_SIZE_T;
-        auto variableIndex = MAX_SIZE_T;
+            std::map<VariablePtr, std::vector<VariablePtr>> primaryExternalVariables;
```

Comment:
Should this be a declared type instead of a combo of std containers?

Reply:
This is an internal type which is only used here. So, I wouldn't bother with creating a type for it.